### ISSUE TITLE
mergerfs 2.30 on stretch armhf - Unhandled fault: alignment fault (0x92000021) at 0x00000000f4801d3b

### ISSUE THREAD

**Issue by vincent-ogury** [trapexit/mergerfs#803]
**Describe the bug**

I try to install the latest mergerfs on a vero4k. For now the only available version in the repos is the 2.18.0.
The vero4k is on stretch on an armhf architecture.

When I install https://github.com/trapexit/mergerfs/releases/download/2.30.0/mergerfs_2.30.0.debian-stretch_armhf.deb
And try to mount my media like I do on the 2.18.0, a simple "ls" break.

**To Reproduce**
Steps to reproduce the behavior. List **all** steps to reproduce. All settings.

- install mergerfs (2.18) from debian/stretch on armhf
- install the package from https://github.com/trapexit/mergerfs/releases/download/2.30.0/mergerfs_2.30.0.debian-stretch_armhf.deb
- mount my media:
```
sudo mergerfs -o defaults,allow_other,category.create=mfs,use_ino -d /media/\*/media /mnt/media
```
- ls /mnt/media

**Expected behavior**
The mergerfs should not crash at all. The package is for jessie armhf, so I except this distribution is supported.

**System information:**
 - OS, kernel version: `uname -a`
Linux vero4kp 3.14.29-158-osmc #1 SMP Thu Apr 2 18:40:24 UTC 2020 aarch64 GNU/Linux

 - mergerfs version: `mergerfs -V`

normal version:
mergerfs version: 2.18.0
FUSE library version: 2.9.7
fusermount version: 2.9.7
using FUSE kernel interface version 7.19

updated version:
mergerfs version: 2.30.0
FUSE library version: 2.9.7-mergerfs_2.30.0
fusermount version: 2.9.7-mergerfs_2.30.0
using FUSE kernel interface version 7.31

 - mergerfs settings
defaults,allow_other,category.create=mfs,use_ino

 - List of drives, filesystems, & sizes: `df -h`
```
Filesystem           Size  Used Avail Use% Mounted on
devtmpfs             757M     0  757M   0% /dev
tmpfs                879M   28M  851M   4% /run
/dev/vero-nand/root   14G  5.1G  8.1G  39% /
tmpfs                879M  4.0K  879M   1% /dev/shm
tmpfs                5.0M     0  5.0M   0% /run/lock
tmpfs                879M     0  879M   0% /sys/fs/cgroup
tmpfs                176M     0  176M   0% /run/user/1000
/dev/sdb1            916G  133G  738G  16% /media/ext2
/dev/sdc1            932G  801G  131G  86% /media/WD1042-1TB-a
/dev/sdd1            932G  828G  104G  89% /media/TB2017-2TB-a
/dev/sdd2            932G  692G  240G  75% /media/TB2017-2TB-b
/dev/sda1            932G  5.0G  927G   1% /media/WD10JM-1TB-a
backup               3.7T  2.3T  1.4T  63% /mnt/backup
media                4.6T  2.5T  2.1T  54% /mnt/media
```

 - A strace of the application having a problem: `strace -f -o /tmp/app.strace.txt <cmd>` or `strace -f -p <appPID> -o /tmp/app.strace.txt`
none. I simplify do a "ls"

 - strace of mergerfs while app tried to do it's thing: `strace -f -p <mergerfsPID> -o /tmp/mergerfs.strace.txt`
[strace.txt](https://github.com/trapexit/mergerfs/files/5033761/strace.txt)

**Additional context**

**Comment by trapexit** [comment]
Why did you install the Jessie package on a Stretch setup?

**Comment by vincent-ogury** [comment]
sorry wrong link. I fix my comment. I confirm I take the stretch link above.

I have success on tmpfs volume, but each time I try my external hard drive something goes wrong. No issue on 2.18.0.

**Comment by trapexit** [comment]
I'm testing in a qemu environment and it seems to work fine. I have an old RPi2 laying around somewhere that I can try setting up to test.

Have you tried building it yourself?

From what I read that system SoC is 64bit. I'm guessing the installed OS is 32bit?

**Comment by vincent-ogury** [comment]
This is OSMC. most of the package use armhf (32 bits I guess), and the kernel is arm64.

```
dpkg -l | awk '{print $4}' | sort | uniq -c
     97 all
      2 arm64
    527 armhf

dpkg -l | grep arm64
ii  vero364-image-3.14.29-158-osmc:arm64 158                               arm64        Linux kernel binary image for version 3.14.29-158-osmc
ii  vero364-kernel-osmc:arm64            3.9.158                           arm64        Kernel meta package bringing in the latest OSMC kernel for this device
```

May be I should try a 64bits version?

**Comment by trapexit** [comment]
Worth a try.

**Comment by vincent-ogury** [comment]
I can't:
```
sudo dpkg -i /home/osmc/src/mergerfs_2.30.0.debian-stretch_arm64.deb

(Reading database ... 38482 files and directories currently installed.)
Preparing to unpack .../mergerfs_2.30.0.debian-stretch_arm64.deb ...
Unpacking mergerfs:arm64 (2.30.0~debian-stretch) over (2.18.0-1) ...
dpkg: dependency problems prevent configuration of mergerfs:arm64:
 mergerfs:arm64 depends on libc6 (>= 2.17).
 mergerfs:arm64 depends on libgcc1 (>= 1:3.0).
 mergerfs:arm64 depends on libstdc++6 (>= 5.2).
 mergerfs:arm64 depends on fuse.

dpkg: error processing package mergerfs:arm64 (--install):
 dependency problems - leaving unconfigured
Processing triggers for man-db (2.7.6.1-2) ...
Errors were encountered while processing:
 mergerfs:arm64
```

**Comment by vincent-ogury** [comment]
I have try to compile the 2.30.0 myself.

I try make with:

- no option
- STATIC=1
- LTO=1
- USE_XATTR=0
...

always the same error.

I try to build tags in reverse order to see when it works.

**Comment by vincent-ogury** [comment]
This version works:
```
mergerfs version: 2.24.0
FUSE library version: 2.9.8-mergerfs
fusermount version: 2.9.7
using FUSE kernel interface version 7.19
```

any above failed the same way.

**Comment by vincent-ogury** [comment]
maybe it is related to the usage of internal fuse. I test without.

**Comment by vincent-ogury** [comment]
I try the 2.25.0 with:
```
make INTERNAL_FUSE=0
```

same error (with internal fuse or external fuse).

so something in this version has break compatibility

**Comment by trapexit** [comment]
People have used mergerfs on debian stretch armhf for a while. It's unlikely that no one else ran into this over the past 2+ years.

I just installed OSMC on a RPi2, installed the 2.30.0 debian stretch armhf deb, plugged in a USB stick with a vfat filesystem, used mergerfs to pool that usb stick and /home/osmc with the options you have... works fine. 

Is there anything more you might be doing in terms of the setup?

**Comment by trapexit** [comment]
What version of OSMC do you have installed on the device? I used the latest which you don't appear to be using.

**Comment by vincent-ogury** [comment]
I have a vero4k+. This is a device from OSMC. I use the last release.

I tag the tag 2.25.0 and do this until it works:

```
git reset --hard HEAD^
```

This commit make the crash:

* 3631ab7  optimize readdir file dedup  (1 year, 9 months ago, Antonio SJ Musumeci)

To test I do:

```
git clean -fdx
make -j4
./mergerfs -V; ./mergerfs -o defaults,use_ino -d /media/\*/media /tmp/test; sudo umount /tmp/test
```

And I try the
```
ln /tmp/test
```

**Comment by vincent-ogury** [comment]
Can I activate something for debugging ? with a maximum details ?
at build ?
then I rerun a strace or something so you can may be spot the problem with my disk ?

I have 3 usb disks of 1TB, 1 of 2TB, 2 of 5TB, Western Digital and Toshiba.

I've got UTF8 encoding filename (maybe bad encoding too) on the disk. 
The root dir contains simple names:

```
Enfants  Mangas  Movies  New  Tv Series  Vu
```

**Comment by trapexit** [comment]
According to what you provided it's a data alignment issue. No debugging mode would really help narrow it down specifically. You'd need to build it with debugging enabled `make DEBUG=1` and then run it in the debugger.

As for encoding... that shouldn't ever be a thing as in POSIX the filename is just a char array. Interpretation is done elsewhere.

**Comment by vincent-ogury** [comment]
OSMC for vero4k devices should move to debian/buster soon. maybe september or end of august.
It should be better afterward.

This system is weird, 64bits kernel with everything in 32bits. I don't understand why. Maybe due to close drivers for their devices.

**Comment by vincent-ogury** [comment]
For the debugger, I run strace?

**Comment by trapexit** [comment]
No. gdb.

```
# gdb ./mergerfs 
run -f -o defaults,use_ino -d /media/\*/media /tmp/test

# when it crashes

thread apply all bt
```

**Comment by vincent-ogury** [comment]
I take back the version 2.30.0.
I confirm it looks like in the fasthash patch:

```
osmc@vero4kp:~/src/mergerfs$ gdb ./build/mergerfs
GNU gdb (Debian 7.12-6) 7.12.0.20161007-git
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "arm-linux-gnueabihf".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./build/mergerfs...done.
(gdb) run -f -o defaults,use_ino -d /media/\*/media /tmp/test
Starting program: /home/osmc/src/mergerfs/build/mergerfs -f -o defaults,use_ino -d /media/\*/media /tmp/test
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/arm-linux-gnueabihf/libthread_db.so.1".
FUSE library version: 2.9.7-mergerfs_2.30.0
[New Thread 0xf705ec40 (LWP 13824)]
unique: 1, opcode: INIT (26), nodeid: 0, insize: 56, pid: 0
INIT: 7.22
[New Thread 0xf675bc40 (LWP 13825)]
flags=0x0000f7fb
max_readahead=0x00020000
   INIT: 7.31
   flags=0x00008079
[New Thread 0xf5cf4c40 (LWP 13826)]
   max_readahead=0x00020000
   max_write=0x00100000
   max_background=0
   congestion_threshold=0
   max_pages=32
   unique: 1, success, outsize: 40
[New Thread 0xf53f1c40 (LWP 13827)]
unique: 2, opcode: GETATTR (3), nodeid: 1, insize: 56, pid: 13828
getattr /
   unique: 2, success, outsize: 120
unique: 3, opcode: OPENDIR (27), nodeid: 1, insize: 48, pid: 13828
opendir flags: 0x24800 /
   opendir[-169834664] flags: 0x24800 /
   unique: 3, success, outsize: 32
unique: 4, opcode: READDIR (28), nodeid: 1, insize: 80, pid: 13828

Thread 5 "mergerfs" received signal SIGBUS, Bus error.
[Switching to Thread 0xf53f1c40 (LWP 13827)]
0xab0d10be in fasthash64 (buf=0xf4800643, len=9, seed=8390876191111276916) at src/fasthash.cpp:48
48          v  = *pos++;
```

**Comment by vincent-ogury** [comment]
```
(gdb) backtrace
#0  0xab0d10be in fasthash64 (buf=0xf4800643, len=9, seed=8390876191111276916) at src/fasthash.cpp:48
#1  0xab0c977c in HashSet::put (this=0xf53f135c, str_=0xf4800643 "Tv Series", len_=9) at src/hashset.hpp:48
#2  0xab0c95d6 in l::readdir (branches_=..., dirname_=0xf5e08760 "/", buf_=0xf5e00510) at src/fuse_readdir_posix.cpp:91
#3  0xab0c971e in FUSE::readdir_posix (branches_=..., dirname_=0xf5e08760 "/", buf_=0xf5e00510)
    at src/fuse_readdir_posix.cpp:121
#4  0xab0df100 in FUSE::readdir (ffi_=0xf53f1498, buf_=0xf5e00510) at src/fuse_readdir.cpp:47
#5  0xab0f2262 in fuse_fs_readdir (fs=0xab12c0f0, fi=0xf53f1498, buf=0xf5e00510) at lib/fuse.c:1906
#6  0xab0f4d22 in readdir_fill (f_=0xab12c818, req_=0xf4800470, d_=0xf5e00510, fi_=0xf53f1498) at lib/fuse.c:3366
#7  0xab0f4e6c in fuse_lib_readdir (req_=0xf4800470, ino_=1, size_=4096, off_=0, llffi_=0xf53f14f8) at lib/fuse.c:3432
#8  0xab0f96d2 in do_readdir (req=0xf4800470, nodeid=1, inarg=0xf53fb030) at lib/fuse_lowlevel.c:1392
#9  0xab0fb4f0 in fuse_ll_process_buf (data=0xab12cea8, buf=0xf53f1620, ch=0xab12af48) at lib/fuse_lowlevel.c:2539
#10 0xab0fcd58 in fuse_session_process_buf (se=0xab12c548, buf=0xf53f1620, ch=0xab12af48) at lib/fuse_session.c:87
#11 0xab0f718a in fuse_do_work (data=0xab12cb28) at lib/fuse_loop_mt.c:93
#12 0xf72685d8 in start_thread () from /lib/arm-linux-gnueabihf/libpthread.so.0
#13 0xf72096fa in ?? () from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

**Comment by vincent-ogury** [comment]
Sorry, I'm not a c++ dev. if anything I can do to send you a useful output.

**Comment by vincent-ogury** [comment]
I try every value for `inodecalc`, none of them works. same issue.

**Comment by vincent-ogury** [comment]
I forget the thread apply command. Here the full stack again:

```
osmc@vero4kp:~/src/mergerfs$ gdb ./build/mergerfs
GNU gdb (Debian 7.12-6) 7.12.0.20161007-git
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "arm-linux-gnueabihf".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./build/mergerfs...done.
(gdb) run -f -o defaults,use_ino -d /media/\*/media /tmp/test
Starting program: /home/osmc/src/mergerfs/build/mergerfs -f -o defaults,use_ino -d /media/\*/media /tmp/test
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/arm-linux-gnueabihf/libthread_db.so.1".
FUSE library version: 2.9.7-mergerfs_2.30.0
[New Thread 0xf7382c40 (LWP 15402)]
unique: 1, opcode: INIT (26), nodeid: 0, insize: 56, pid: 0
INIT: 7.22
flags=0x0000f7fb
max_readahead=0x00020000
   INIT: 7.31
   flags=0x00008079
   max_readahead=0x00020000
   max_write=0x00100000
   max_background=0
   congestion_threshold=0
   max_pages=32
   unique: 1, success, outsize: 40
[New Thread 0xf6a7fc40 (LWP 15403)]
[New Thread 0xf5ff4c40 (LWP 15404)]
[New Thread 0xf56f1c40 (LWP 15405)]
unique: 2, opcode: GETATTR (3), nodeid: 1, insize: 56, pid: 15406
getattr /
   unique: 2, success, outsize: 120
unique: 3, opcode: OPENDIR (27), nodeid: 1, insize: 48, pid: 15406
opendir flags: 0x24800 /
   opendir[-187660456] flags: 0x24800 /
   unique: 3, success, outsize: 32
unique: 4, opcode: READDIR (28), nodeid: 1, insize: 80, pid: 15406

Thread 3 "mergerfs" received signal SIGBUS, Bus error.
[Switching to Thread 0xf6a7fc40 (LWP 15403)]
0xaac300be in fasthash64 (buf=0xf4b00643, len=9, seed=8390876191111276916) at src/fasthash.cpp:48
48	    v  = *pos++;
(gdb) thread apply all bt

Thread 5 (Thread 0xf56f1c40 (LWP 15405)):
#0  0xf759349e in read () from /lib/arm-linux-gnueabihf/libpthread.so.0
#1  0x00000000 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 4 (Thread 0xf5ff4c40 (LWP 15404)):
#0  0xf759349e in read () from /lib/arm-linux-gnueabihf/libpthread.so.0
#1  0x00000000 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 3 (Thread 0xf6a7fc40 (LWP 15403)):
#0  0xaac300be in fasthash64 (buf=0xf4b00643, len=9, seed=8390876191111276916) at src/fasthash.cpp:48
#1  0xaac2877c in HashSet::put (this=0xf6a7f35c, str_=0xf4b00643 "Tv Series", len_=9) at src/hashset.hpp:48
#2  0xaac285d6 in l::readdir (branches_=..., dirname_=0xf4d08760 "/", buf_=0xf4d00510) at src/fuse_readdir_posix.cpp:91
#3  0xaac2871e in FUSE::readdir_posix (branches_=..., dirname_=0xf4d08760 "/", buf_=0xf4d00510)
    at src/fuse_readdir_posix.cpp:121
#4  0xaac3e100 in FUSE::readdir (ffi_=0xf6a7f498, buf_=0xf4d00510) at src/fuse_readdir.cpp:47
#5  0xaac51262 in fuse_fs_readdir (fs=0xaac8b0f0, fi=0xf6a7f498, buf=0xf4d00510) at lib/fuse.c:1906
#6  0xaac53d22 in readdir_fill (f_=0xaac8b818, req_=0xf4b00470, d_=0xf4d00510, fi_=0xf6a7f498) at lib/fuse.c:3366
#7  0xaac53e6c in fuse_lib_readdir (req_=0xf4b00470, ino_=1, size_=4096, off_=0, llffi_=0xf6a7f4f8) at lib/fuse.c:3432
#8  0xaac586d2 in do_readdir (req=0xf4b00470, nodeid=1, inarg=0xf6a89030) at lib/fuse_lowlevel.c:1392
#9  0xaac5a4f0 in fuse_ll_process_buf (data=0xaac8bea8, buf=0xf6a7f620, ch=0xaac89f48) at lib/fuse_lowlevel.c:2539
#10 0xaac5bd58 in fuse_session_process_buf (se=0xaac8b548, buf=0xf6a7f620, ch=0xaac89f48) at lib/fuse_session.c:87
#11 0xaac5618a in fuse_do_work (data=0xaac8b7e0) at lib/fuse_loop_mt.c:93
#12 0xf758c5d8 in start_thread () from /lib/arm-linux-gnueabihf/libpthread.so.0
#13 0xf752d6fa in ?? () from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 2 (Thread 0xf7382c40 (LWP 15402)):
#0  0xf759349e in read () from /lib/arm-linux-gnueabihf/libpthread.so.0
#1  0x00000000 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 1 (Thread 0xf748e000 (LWP 15397)):
#0  0xf7595384 in __libc_do_syscall () from /lib/arm-linux-gnueabihf/libpthread.so.0
#1  0xf7592744 in do_futex_wait.constprop () from /lib/arm-linux-gnueabihf/libpthread.so.0
#2  0xf7592826 in __new_sem_wait_slow.constprop.0 () from /lib/arm-linux-gnueabihf/libpthread.so.0
#3  0xaac564d2 in fuse_session_loop_mt (se=0xaac8b548, _threads=4) at lib/fuse_loop_mt.c:210
#4  0xaac5f0f4 in fuse_loop_mt (f=0xaac8b818) at lib/fuse_mt.c:118
#5  0xaac5c7fc in fuse_main_common (argc=14, argv=0xaac8b3e0, op=0xfffef034, op_size=200, user_data=0xfffef108, compat=0)
    at lib/helper.c:386
#6  0xaac5c84a in fuse_main_real (argc=14, argv=0xaac8b3e0, op=0xfffef034, op_size=200, user_data=0xfffef108)
    at lib/helper.c:400
#7  0xaac2468c in l::main (argc_=7, argv_=0xfffef594) at src/mergerfs.cpp:177
#8  0xaac246ea in main (argc=7, argv=0xfffef594) at src/mergerfs.cpp:188
(gdb)
```

**Comment by trapexit** [comment]
Yes because it's nothing to do with inodecalc. It's the hashing function (which is used for other purposes) that the CPU doesn't like.

Can you, from the first stack trace, provide me values of some variables? from gdb?

``` gdb

p pos
p buf
p end
```

**Comment by vincent-ogury** [comment]
```
(gdb) p pos
$4 = (const uint64_t *) 0xf4b0064b
(gdb) p buf
$5 = (const void *) 0xf4b00643
(gdb) p end
$6 = (const uint64_t *) 0xf4b0064b
```

**Comment by vincent-ogury** [comment]
It seems to be the "Tv Series" directory who cause the issue.
When I try to mount disk without it, the "ls" works.
If I try to mount only one disk with it failed.

**Comment by vincent-ogury** [comment]
I try to mount the directory above, it failed with another one:
```
#1  0xaac2a77c in HashSet::put (this=0xf5ef435c, str_=0xf4a086a3 ".DS_Store", len_=9) at src/hashset.hpp:48
```

**Comment by trapexit** [comment]
You would have to print the buf pointer value for each value to really compare. It's an alignment issue. Meaning the CPU doesn't like where in memory that data is when interpreted in a certain way. 0xf4b00643 is not on a 4 byte or 8 byte boundary.

I'll have to see what I can do to manage the alignment issue.

**Comment by vincent-ogury** [comment]
I put here the cpuinfo if that can help:

```
osmc@vero4kp:~/src/mergerfs/src$ cat /proc/cpuinfo
Processor       : AArch64 Processor rev 4 (aarch64)
processor       : 0
processor       : 1
processor       : 2
processor       : 3
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 wp half thumb fastmult vfp edsp neon vfpv3 tlsi vfpv4 idiva idivt
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 4

Hardware        : Vero4KPlus
Serial          : 210d04007da66121600fd86ae7e3543c
```

**Comment by vincent-ogury** [comment]
I have found a patch for my case:

https://github.com/rurban/smhasher

I switch to farmhash, which seems 2x faster.

[p.diff.zip](https://github.com/trapexit/mergerfs/files/5036607/p.diff.zip)

**Comment by vincent-ogury** [comment]
ls / create works. can you give me your filling about that?

**Comment by trapexit** [comment]
I'm looking at different hashes or reworking fasthash. Need to compare biases, etc.

**Comment by vincent-ogury** [comment]
ok. while you check, I'm testing with farmhash on my system. I think my patch is ok, but well, I'm not a cpp dev.
What's could be wrong if the hash is not good?
What are the possible issues?

**Comment by trapexit** [comment]
Collisions. This is used to dedup filenames. An invalid collision will invalidly remove a file from listing properly. It's extremely unlikely. It was a way to improve the speed and memory usage of deduping readdirs.

**Comment by vincent-ogury** [comment]
Ok, well, I will be happy to test an approved patch as soon as you get one ;)

On the site I sent, the guy has compare ton of hash, and the summary indicate:

```
So the fastest hash functions on x86_64 without quality problems are:

xxh3low
wyhash
t1ha2_atonce
FarmHash (not portable, too machine specific: 64 vs 32bit, old gcc, ...)
Spooky32
fasthash32
MUM/mir (different results on 32/64-bit archs)
```

**Comment by trapexit** [comment]
Yeah, I saw. I'm looking over the docs now. I just need time to compare.

**Comment by vincent-ogury** [comment]
👍 

wyhash seems even faster. and it is only 1 .h file, with inline function.

Ok I let you do.

Thanks a lot for the help. I'm gonna use this wonderful tools you got for my media and backups.

**Comment by trapexit** [comment]
https://github.com/trapexit/mergerfs/tree/wyhash

Does that work on your system?

**Comment by celogeek** [comment]
I don't know. I have test farmhash. This one works. 
I can try.

**Comment by trapexit** [comment]
Sorry... I was asking you to test that branch.

**Comment by celogeek** [comment]
Yes. It works. It is a bit faster than farmhash it seems. 

```

osmc@vero4kp:~/src/mergerfs$ time find /tmp/test -ls > /dev/null; time find /mnt/media -ls > /dev/null

real    0m0.378s
user    0m0.110s
sys     0m0.050s

real    0m0.449s
user    0m0.120s
sys     0m0.040s
```

Mount options:
```
./build/mergerfs -o defaults,use_ino,cache.files=off,dropcacheonclose=true,category.create=mfs /media/\*/media /tmp/test
```

**Comment by trapexit** [comment]
I really wouldn't expect much performance difference given the overhead of all that but every little bit counts I guess.

I just ran a check and of 1098892 unique filenames on my system I got zero collisions so I'm good with this change. 

If you don't mind I'll release this in the next minor release since this has been around for years and you're the only person to notice. Want to see if anything else crops up from the new release. Probably won't be more than a week or two.
